stdlib/public/Concurrency/TaskStatus.cpp

```diff
@@ -576,58 +550,79 @@ static void performEscalationAction(TaskStatusRecord *record,
-    if (status.getStoredPriority() >= newPriority)
-      return;
-    status = status.withEscalatedPriority(newPriority);
 
-    // TODO: attempt to escalate the thread running the task, if it's
-    // currently running.  This probably requires the task to be enqueued
-    // on a standard executor.
+  auto oldStatus = task->_private().Status.load(std::memory_order_relaxed);
+  auto newStatus = oldStatus;
 
-    // Perform escalation operations for all the status records.
-    for (auto cur: status.records()) {
-      performEscalationAction(cur, newPriority);
+  while (true) {
+    // Fast path: check that the stored priority is already at least
+    // as high as the desired priority.
+    if (oldStatus.getStoredPriority() >= newPriority)
+      return oldStatus.getStoredPriority();
+
+    // Regardless of whether status record is locked or not, update the priority
+    // and RO bit on the task status
+    if (oldStatus.isRunning()) {
+      newStatus = oldStatus.withEscalatedPriority(newPriority);
+    } else {
+      newStatus = oldStatus.withNewPriority(newPriority);
     }
-  });
-
-  return status.getStoredPriority();
-}
-
-void AsyncTask::flagAsRunning_slow() {
-  withStatusRecordLock(this, LockContext::OnTask,
-                       [&](ActiveTaskStatus &status) {
-    assert(!status.isRunning());
 
-    status = status.withRunning(true);
-    if (status.isStoredPriorityEscalated()) {
-      status = status.withoutStoredPriorityEscalation();
-      Flags.setPriority(status.getStoredPriority());
-      concurrency::trace::task_flags_changed(this, Flags.getOpaqueValue());
+    if (task->_private().Status.compare_exchange_weak(oldStatus, newStatus,
+            /* success */ std::memory_order_relaxed,
+            /* failure */ std::memory_order_relaxed)) {
+      break;
     }
-  });
-}
-
-void AsyncTask::flagAsSuspended_slow() {
-  withStatusRecordLock(this, LockContext::OnTask,
-                       [&](ActiveTaskStatus &status) {
-    assert(status.isRunning());
+  }
 
-    status = status.withRunning(false);
-    if (status.isStoredPriorityEscalated()) {
-      status = status.withoutStoredPriorityEscalation();
-      Flags.setPriority(status.getStoredPriority());
-      concurrency::trace::task_flags_changed(this, Flags.getOpaqueValue());
+#if SWIFT_CONCURRENCY_ENABLE_PRIORITY_ESCALATION
+  if (newStatus.isRunning()) {
+    // The task is running, escalate the thread that is running it. The
+    // dispatch SPI does an atomic debounce on the lock address to make sure
+    // that the thread that is expected to be the drainer is still the
+    //
+    // The current dispatch API only looks at the owner field but
+    // eventually, we might want to have a fancy debounce logic on the
+    // full 64 bit states including flags.
+    uint64_t *drain_lock_and_flags = (uint64_t *)&task->_private().Status;
+    dispatch_lock_t *drain_lock = (dispatch_lock_t *)((char *) drain_lock_and_flags + sizeof(uint32_t));
```

Comment:
`offsetof` or casting to `ActiveTaskStatus *` and taking the address of the field would be a bit clearer here.